harness/determined/searcher/core_search_runner.py

```diff
@@ -0,0 +1,121 @@
+    determinedexperimentv1State,
+    get_GetExperiment,
+    post_PauseExperiment,
+)
+from determined.experimental import client
+from determined.searcher.search_method import Operation, SearchMethod
+from determined.searcher.search_runner import SearchRunner
+
+
+class CoreSearchRunner(SearchRunner):
+    def __init__(self, search_method: SearchMethod, context: det.core.Context) -> None:
+        super().__init__(search_method)
+        self.context = context
+        self.info = det.get_cluster_info()
+
+        assert self.info is not None, "CoreSearchRunner only runs on-cluster"
+        self.latest_checkpoint = self.info.latest_checkpoint
+
+    def run(
+        self,
+        exp_config: Dict[str, Any],
+        context_dir: Optional[str] = None,
+    ) -> int:
+        logging.info("CoreSearchRunner.run")
+        client._require_singleton(lambda: None)()
+
+        operations: Optional[List[Operation]] = None
+
+        if context_dir is None:
+            context_dir = os.getcwd()
+
+        if self.latest_checkpoint is not None:
+            experiment_id, operations = self.load_state(self.latest_checkpoint)
+            logging.info(f"Resuming HP searcher for experiment {experiment_id}")
+        else:
+            logging.info("No latest checkpoint. Starting new experiment.")
+            exp = client.create_experiment(exp_config, context_dir)
+            self.search_method.searcher_state.experiment_id = exp.id
+            self.search_method.searcher_state.last_event_id = 0
+            self.save_state(exp.id, [])
+            experiment_id = exp.id
+
+        client._require_singleton(lambda: None)()
+        assert client._determined is not None
+        session = client._determined._session
+        self.run_experiment(experiment_id, session, operations)
+
+        return experiment_id
+
+    def load_state(self, storage_id: str) -> Tuple[int, List[Operation]]:
+        with self.context.checkpoint.restore_path(storage_id) as path:
+            experiment_id = self.search_method.load(path)
+            with path.joinpath("ops").open("rb") as ops_file:
+                operations = pickle.load(ops_file)
+            return experiment_id, operations
+
+    def save_state(self, experiment_id: int, operations: List[Operation]) -> None:
+        steps_completed = self.search_method.searcher_state.last_event_id
+        metadata = {"steps_completed": steps_completed}
+        with self.context.checkpoint.store_path(metadata) as (path, storage_id):
+            self.search_method.save(path, experiment_id=experiment_id)
+            with path.joinpath("ops").open("wb") as ops_file:
+                pickle.dump(operations, ops_file)
+
+    def maybe_preempt(self, session: client.Session, multitrial_experiment_id: int) -> bool:
+        if self.context.preempt.should_preempt():
+            # if searcher is preempted, then pause multitrial experiment
+            self._pause_and_wait(session, multitrial_experiment_id)
```

Comment:
Question: how do we become unpreempted?

IIUC, we can get preempted either by scheduler decision or user action.

In either case, it looks like we pause the experiment.  So if we got preemtped by the scheduler, we're going to push the experiment into a PAUSED state, then we put ourselves into a PAUSED state, then we won't ever wake back up.

Reply:
@aciborowska , would you mind looking at this one?